File: javascript/Problem24.js
/*
 * A permutation is an ordered arrangement of objects. For example, 3124 is one
 * possible permutation of the digits 1, 2, 3 and 4. If all of the permutations
 * are listed numerically or alphabetically, we call it lexicographic order. The
 * lexicographic permutations of 0, 1 and 2 are:
 * 
 * 012   021   102   120   201   210
 * 
 * What is the millionth lexicographic permutation of the digits 0, 1, 2, 3, 4, 5,
 * 6, 7, 8 and 9?
 */

function factorial(x) {
  return x < 2 ? 1 : x * factorial(x - 1)
}

function permutationAt(xs, n) {
  if (xs.length < 2) {
    return xs
  }
  const m = factorial(xs.length - 1)
  const index = Math.floor(n / m)
  const ys = xs.slice(0, index).concat(xs.slice(index + 1))
  return [xs[index]].concat(permutationAt(ys, n - m * index))
}

const answer = permutationAt([0, 1, 2, 3, 4, 5, 6, 7, 8, 9], 999999).join('')
console.log(answer)

File: go/Problem24.go
/*
 * A permutation is an ordered arrangement of objects. For example, 3124 is one
 * possible permutation of the digits 1, 2, 3 and 4. If all of the permutations
 * are listed numerically or alphabetically, we call it lexicographic order. The
 * lexicographic permutations of 0, 1 and 2 are:
 *
 * 012   021   102   120   201   210
 *
 * What is the millionth lexicographic permutation of the digits 0, 1, 2, 3, 4, 5,
 * 6, 7, 8 and 9?
 */
package main

import (
	"fmt"
	"math"
)

func Factorial(x int) (result int) {
	for result = 1; x > 0; x-- {
		result *= x
	}
	return
}

func PermutationAt(xs []int, n int) {
	if len(xs) < 2 {
		return
	}
	m := Factorial(len(xs) - 1)
	index := n / m
	x := xs[index]
	for i := index; i > 0; i-- {
		xs[i] = xs[i-1]
	}
	xs[0] = x
	PermutationAt(xs[1:], n-m*index)
}

func main() {
	digits := []int{0, 1, 2, 3, 4, 5, 6, 7, 8, 9}
	PermutationAt(digits, 999999)
	answer := 0
	for i, digit := range digits {
		answer += digit * int(math.Pow10(len(digits)-i-1))
	}
	fmt.Println(answer)
}
